subprojects/base-services/src/main/java/org/gradle/internal/service/DefaultServiceRegistry.java

```diff
@@ -524,19 +524,28 @@ public ServiceProvider getService(LookupContext context, TypeSpec serviceType) {
             if (providers == null) {
                 return null;
             }
-            List<ServiceProvider> candidates = new ArrayList<ServiceProvider>();
+            ServiceProvider singleCandidate = null;
+            List<ServiceProvider> candidates = null;
             for (Provider provider : providers) {
                 ServiceProvider service = provider.getService(context, serviceType);
                 if (service != null) {
-                    candidates.add(service);
+                    if (singleCandidate == null) {
+                        singleCandidate = service;
+                    } else {
+                        if (candidates == null) {
+                            candidates = new ArrayList<ServiceProvider>(2);
+                            candidates.add(singleCandidate);
+                        }
+                        candidates.add(service);
+                    }
                 }
             }
 
-            if (candidates.size() == 0) {
+            if (candidates == null && singleCandidate == null) {
```

Comment:
Ok I mixed 2 things, but it doesn't change the reasoning. Before, we created an internal `ArrayList` that would contain all candidate services. Knowing that in practice, if there are more than one, then it's an error case. So, we were creating an array list, each time, for each service, knowing that in practice it would contain a single entry. With the change, we defer the creation of the list to the point where we know there will be an error. We could use a wrapping object like `FastSet` to implement this but it would involve the creation of an object we don't need.